### PR TITLE
AP-9751 # Added submission timestamp conditional logic

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **Medium Risk**
> Payment and scheduling event detection now depend on submission timestamps and shared date parsing; incorrect timestamps or parsing could change when fees or booking flows trigger. SDK major bump may introduce broader conditional-logic behavior changes beyond this repo.
> 
> **Overview**
> Upgrades **`@oneblink/sdk-core`** to **`^10.0.0-beta.2`** so conditional logic (including rules that depend on **submission time**) is evaluated in the SDK instead of partly in the React app.
> 
> **Date helpers for the SDK:** Adds **`parseDate`** and **`addDaysToDate`** in `localisation-service` (date-only `YYYY-MM-DD` at local midnight; full ISO datetimes keep time-of-day) and passes them with **`submissionTimestamp`** into **`checkForPaymentEvent`** / **`checkForSchedulingEvent`** from submit, payment, and scheduling flows.
> 
> **Conditional logic hook:** **`useConditionalLogic`** now passes **`enableSubmission`** into **`generateFormElementsConditionallyShown`** and uses the returned **`isSubmissionEnabled`** instead of re-evaluating submit-enable predicates locally (removes **`cleanFormSubmissionModel`** usage there).
> 
> **Other:** Tightens calculation element typing (`ExpressionParser` generics, numeric-only results) and adds **`morph-expressions`** module declarations; adds vitest coverage for **`parseDate`** / **`generateDate`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 09d16db74602e7ce6ac910c779c5534b664a58e0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/hooks/useConditionalLogic.ts [hunk 1]
```diff
@@ -52,6 +56,9 @@ export default function useConditionalLogic({
       conditionalPredicates,
       formElements,
       submission: model,
+      submissionTimestamp: new Date().toISOString(),
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/hooks/useConditionalLogic.ts]: Fixed by moving conditionally enabled into sdk-core

File: src/hooks/useConditionalLogic.ts [hunk 2]
```diff
@@ -52,6 +56,9 @@ export default function useConditionalLogic({
...
+      parseDate,
+      addDaysToDate,
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/hooks/useConditionalLogic.ts]: Fixed by moving conditionally enabled into sdk-core

File: src/apps/payment-service.ts [hunk 3]
```diff
@@ -125,10 +129,13 @@ export function checkForPaymentSubmissionEvent(formSubmission: FormSubmission):
...
-    formSubmission.definition,
-    formSubmission.submission,
-  )
+  const result = paymentService.checkForPaymentEvent({
+    definition: formSubmission.definition,
+    submission: formSubmission.submission,
+    submissionTimestamp: new Date().toISOString(),
+    parseDate,
+    addDaysToDate,
+  })
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/apps/payment-service.ts:141]: Fixed by passing the completion timestamp before submission and submission timestamp after submission

File: src/apps/localisation-service.ts [hunk 4]
```diff
@@ -309,6 +309,26 @@ function generateDateOffset({
...
+ * Parse a date/datetime string for conditional logic evaluation. Date-only
+ * values (`YYYY-MM-DD`) are interpreted in the user's local timezone.
+ */
+export function parseDate(value: string): Date {
+  return (
+    generateDate({
+      value,
+      daysOffset: undefined,
+    }) ?? new Date(NaN)
+  )
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/apps/localisation-service.ts:324]: This is not correct for `date-fns` `parse()`. Unit test has been added to confirm.

[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/apps/localisation-service.ts:324]: Already addressed in previous comment

File: src/apps/scheduling-service.ts [hunk 5]
```diff
@@ -30,8 +30,10 @@ async function getPaymentConfiguration({
   }
 
   await setSchedulingBooking(schedulingBooking)
-  const paymentSubmissionEventConfiguration =
-    checkForPaymentSubmissionEvent(formSubmissionResult)
+  const paymentSubmissionEventConfiguration = checkForPaymentSubmissionEvent(
+    formSubmissionResult,
+    formSubmissionResult.submissionTimestamp ?? new Date().toISOString(),
+  )
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/apps/scheduling-service.ts:36]: Already addressed in previous comment

File: src/apps/localisation-service.ts [hunk 6]
```diff
@@ -309,6 +309,26 @@ function generateDateOffset({
...
+/**
+ * Parse a date/datetime string for conditional logic evaluation. Date-only
+ * values (`YYYY-MM-DD`) are interpreted in the user's local timezone.
+ */
+export function parseDate(value: string): Date {
+  return (
+    generateDate({
+      value,
+      daysOffset: undefined,
+    }) ?? new Date(NaN)
```
Comment (divporter) [inline src/apps/localisation-service.ts:323]: Could you add a comment that we are forcing an invalid date plz

**Review by divporter — APPROVED** [review]

**Comment by divporter** [comment]
I'm happy to disregard bug bot if you are.
